File: android/app/src/main/java/com/golfiq/hud/analytics/AnalyticsCoordinator.kt
package com.golfiq.hud.analytics

import android.content.Context
import com.golfiq.hud.model.FeatureFlagConfig
import com.golfiq.hud.telemetry.TelemetryClient
import java.net.URL

class AnalyticsCoordinator(
    private val context: Context,
    private val telemetryClient: TelemetryClient,
    baseUrl: URL,
    private val dsnProvider: () -> String?,
) {
    private val crashBridge = CrashTelemetryBridge(URL(baseUrl, "telemetry"))
    private var lastSignature: String? = null

    fun apply(flags: FeatureFlagConfig, configHash: String?) {
        val dsn = dsnProvider()?.takeIf { !it.isNullOrBlank() }
        AnalyticsInit.initIfEnabled(context, flags, dsn)
        crashBridge.updateFlags(flags)
        val signature = listOf(
            configHash ?: "local",
            flags.analyticsEnabled,
            flags.crashEnabled,
            !dsn.isNullOrEmpty(),
        ).joinToString(":")
        if (signature != lastSignature) {
            telemetryClient.logAnalyticsConfig(
                analyticsEnabled = flags.analyticsEnabled,
                crashEnabled = flags.crashEnabled,
                dsnPresent = !dsn.isNullOrEmpty(),
                configHash = configHash ?: "local",
            )
            lastSignature = signature
        }
    }
}

File: ios/Analytics/AnalyticsCoordinator.swift
import Foundation

final class AnalyticsCoordinator {
    private let telemetry: TelemetryClient
    private let crashBridge: CrashTelemetryBridge
    private let dsnProvider: () -> String?
    private var lastSignature: String?

    init(telemetry: TelemetryClient, baseURL: URL, dsnProvider: @escaping () -> String?) {
        self.telemetry = telemetry
        self.crashBridge = CrashTelemetryBridge(endpoint: baseURL.appendingPathComponent("telemetry"))
        self.dsnProvider = dsnProvider
    }

    func apply(flags: FeatureFlagConfig, configHash: String?) {
        let dsn = dsnProvider()?.trimmingCharacters(in: .whitespacesAndNewlines)
        AnalyticsInit.initIfEnabled(flags: flags, dsn: dsn)
        crashBridge.update(flags: flags)
        let signature = "\(configHash ?? "local"):\(flags.analyticsEnabled):\(flags.crashEnabled):\(!(dsn?.isEmpty ?? true))"
        if signature != lastSignature {
            telemetry.logAnalyticsConfig(
                analyticsEnabled: flags.analyticsEnabled,
                crashEnabled: flags.crashEnabled,
                dsnPresent: !(dsn?.isEmpty ?? true),
                configHash: configHash ?? "local"
            )
            lastSignature = signature
        }
    }
}
